Integrate my ontology changes into views

mkhoshnam opened the pull request:
This PR introduces modifications to src/semantic_world/views/views.py to extend the ontology support.
Specifically:

Added new ontology-related fields to OntologyView (e.g., affordances, manipulation_properties, state_information, grasp_poses, effects, description, joint_type, joint_sim_name).

Updated relevant classes (Drawer, Cabinet, Fridge, Microwave, Oven, Stove, Sink, Dishwasher, Counter, CoffeeMachine, Hood, etc.) to inherit from or integrate with OntologyView.

Refactored portions of the file to support richer semantic information in the views layer.

Purpose

These changes allow integration of a custom ontology with the semantic_world framework, enabling representation of affordances, manipulation properties, and semantic states for objects like drawers, cabinets, fridges, microwaves, etc.

This will make the ontology more expressive and easier to use in downstream reasoning and planning tasks.

Purpose

These changes allow integration of a custom ontology with the semantic_world framework, enabling representation of affordances, manipulation properties, and semantic states for objects like drawers, cabinets, fridges, microwaves, etc.

This will make the ontology more expressive and easier to use in downstream reasoning and planning tasks.

AbdelrhmanBassiouny commented on src/semantic_world/views/views.py:
```diff
@@ -13,6 +13,21 @@
 
 
+@dataclass
+class OntologyView(View):
```
I think OntologyView doesn't convey what this is about, also I think this class shouldn't be a view since it has no relation to the world by itself.

from how this class looks I think something like `ManipulationInfo` would make more sense, and it could be that it's better to use composition here instead of inheritance like this:
```python
@dataclass
class Drawer(View):
    manipulation_info: ManipulationInfo = field(default_factory=ManipulationInfo)
```

mkhoshnam commented:
Implemented requested changes: replaced OntologyView with ManipulationInfo , updated all relevant views to hold manipulation_info, and added per-field documentation in the ManipulationInfo docstring. Ready for re-review.

tomsch420 commented:
Also make the class more specific. If  it  handles information like this in general, make it an abstract class and create subclasses for specific kinds of ManipulationInformation like this. I think you would benefit from a use case and not just the addition of random fields to all views. Perhaps come see us in the office to discuss what you want to do.

tomsch420 commented:
I think we are talking past each other here. Come to our office tomorrow.

AbdelrhmanBassiouny commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+        raise ValueError(f"Body '{body_name}' not found in MuJoCo model")
+
+    # Prefer PrefixedName if your Body expects it; fall back to plain str if needed.
+    try:
```
the Body takes a prefixed name no need for this try ... except

AbdelrhmanBassiouny commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+# ---------- Digital-Twin Controller bound to the semantic view ----------
+
+@dataclass
+class DigitalTwinCabinetAgent:
```
It's very wierd to make a class that is considered an Agent that only controls the cabinet ... in addition sematnic world has APIs that you can use to manipulate the world state and do forward and inverse kinematics, please check the user start guide that was just merged for how to do that.

LucaKro replied:
I also dont get how this is an agent yet, but i may be misunderstanding

AbdelrhmanBassiouny reviewed (changes requested):
I do not think that this is useful to you it seems like a forced integration of semantic world to your work that does not seem to add any meaningful benefits to semantic world or to your work.

To benefit from semantic world you have to use it's features, for example you are trying to control a cabinet, and you did not use semantic world state or joints or degrees of freedom or the ik solver.... instead you are just using mujoco.

So please check the start user guide to be able to make use of semantic world.

Second thing the way you are currently constructing views from ontology/mujoco data were you hope you find matching attributes and if not just set them manually is not a good idea. It would be better if you think about your task and what are the needed attributes that would add a real value to semantic world views and then add them manually to the views because then it becomes useful to everyone not just your application.

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
why "digital_twin_adapter"?

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+Digital-Twin adaptor: bridges semantic_world views to an actual simulator (MuJoCo).
+Use case: open / close a cabinet door by name, using the joint mapped from the semantic view.
+
+Adjust the imports below to match your repo's layout.
```
this also seems copy pasted from an LLM, I dont think we need this here

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+
+# ---------- Build a semantic Cabinet view from a MuJoCo model ----------
+
+def _mk_body(model: mj.MjModel, data: mj.MjData, body_name: str) -> Body:
```
why do we have this? this method basically creates a body if the name is already in the mujoco model? but the object you create doesnt have any information about collision/visuals, and is also not added to the kinematic structure. trying to control a cabinet where none of the bodies are in the kinematic structure does not make any sense

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+    d_body   = _mk_body(model, data, door_body)
+    h_body   = _mk_body(model, data, handle_body)
+
+    handle = _flex_construct(Handle, {"body": h_body})
```
why do we need "_flex_construct"? why cant we just do `Handle(body=h_body)` and `Door(body=d_body, handle=handle)`?

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+        "body": cab_body,
+        "doors": [door],
+        "drawers": [],
+        "joint_type": joint_type,
```
we already have a concept of joints in the semdt, why are you doing it like this instead of creating a connection of the proper kind?

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+
+# ---------- Flexible constructor helper ----------
+
+def _flex_construct(cls, fallback_attrs: dict):
```
i dont like this method exists, I dont think its necessary.

also the name"fallback_attrs" doesnt make any sense to me here

LucaKro commented on src/semantic_world/adapters/digital_twin_adapter.py:
```diff
@@ -0,0 +1,180 @@
+        """
+        Set door openness in [0,1].
+        - hinge: maps to angle between [closed, open], here we assume [0, +90°]
+        - prismatic: maps to linear slide, assume [0, +0.25 m]
```
why do we "assume"? doesnt mujoco joints have jointlimits you can use here?

LucaKro reviewed (changes requested):
while this looks very interesting, I dont think the content of this PR fit with the current vision of semdt yet. to me this looks to me more like the start of its own framework that imports semdt, similar to robokudo, giskard, multiverse etc.
So if i were you I would create my own package that you implement this functionality inside, and import semdt for features you need that we have already implemented. if you notice that some feature is missing in semdt, then you create a PR adding that feature you need.

since you are also working with mujoco there may be some crossover with multiverse, but im not sure, you would need to talk to giang about that.
